### PR TITLE
shutdown workers on failure

### LINKED ISSUE #9390: Ability to release resources associated with a DataLoader
https://github.com/pytorch/pytorch/pull/35795 adds `DataLoader.persistent_workers` which is great from a performance point of view. Unfortunately when running multiple trials using Optuna, DataLoader workers take a long time to shut down after a trial completes.

When running under Windows, I am seeing each worker process allocate a commit size of 5GB. When `num_workers = 4` each trial commits 20GB of memory and it takes idle workers a long time (>1 minute) to shut down. If trials are short, I can easily end up with over 100GB of memory committed by worker processes which results in an out of memory error.

Is it possible to proactively shut down DataLoader workers at the end of a trial?

Does `DataLoader` provide a mechanism to release its resources? Does `teardown()` or other hooks provide a mechanism to close data loaders?

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9390

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/loops/test_loops.py [hunk 1]
```diff
@@ -912,8 +912,9 @@ def val_dataloader(self):
 
 
 @RunIf(min_torch="1.8.0")
-@pytest.mark.parametrize("persistent_workers", (False, True))
-def test_workers_are_shutdown(tmpdir, persistent_workers):
+@pytest.mark.parametrize("should_fail", [False, True])
+@pytest.mark.parametrize("persistent_workers", [False, True])
+def test_workers_are_shutdown(tmpdir, should_fail, persistent_workers):
```
Comment (carmocca) [inline tests/loops/test_loops.py:917]: The test takes 50s to run on my machine. Do you have any ideas how we could improve that?

**Review by carmocca — APPROVED** [review]

File: tests/loops/test_loops.py (hunk 1 above)
Comment (tchaton) [reply tests/loops/test_loops.py:917]: Definitely any issue. I don't have a good idea. Re-creating the processes is quite expensive on Mac.

Comment (cowwoc) [reply tests/loops/test_loops.py:917]: Is it safe to assume that the slowness only occurs if `persistent_workers=False`? Or are you seeing a problem with the value is `True`?

If the problem only occurs on `persistent_workers=False` then I wonder why we haven't run across this problem before as this is the default value of `persistent_workers`.

**Review by rohitgr7 — APPROVED** [review]

File: tests/loops/test_loops.py (hunk 1 above)
Comment (awaelchli) [reply tests/loops/test_loops.py:917]: it has num_workers=1 so yes, no surprise :) 
If we really want to test this, we have to either take the 50s hit or use unittest mock to check the calls as we expect them (and don't explicitly spawn processes).

File: CHANGELOG.md [hunk 2]
```diff
@@ -121,6 +121,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `CombinedLoader` and `max_size_cycle` didn't receive a `DistributedSampler` ([#10374](https://github.com/PyTorchLightning/pytorch-lightning/issues/10374))
 
 
+- Fixed `shutdown workers` on failure ([#10463](https://github.com/PyTorchLightning/pytorch-lightning/issues/10463))
```
Comment (awaelchli) [inline CHANGELOG.md:124]:
```suggestion
- Fixed an issue that prevented the Trainer to shutdown workers when execution is interrupted due to failure([#10463](https://github.com/PyTorchLightning/pytorch-lightning/issues/10463))
```

**Review by awaelchli — APPROVED** [review]

File: tests/loops/test_loops.py (hunk 1 above)
Comment (tchaton) [reply tests/loops/test_loops.py:918]: Done !

**Comment by cowwoc** [comment]
Thank you!

**Pull request merged**